classes/statistics/UsageStatsTotalTemporaryRecordDAO.inc.php

```diff
@@ -0,0 +1,369 @@
+                ON DUPLICATE KEY UPDATE metric = metric_tmp;
+                ';
+        }
+        // load metric
+        DB::statement($metricUpsertSql, [$loadId]);
+    }
+
+    /**
+     * Load total COUNTER submission usage (investigations and requests)
+     */
+    public function loadMetricsCounterSubmissionDaily(string $loadId): void
+    {
+        // construct metric_investigations upsert
+        $metricInvestigationsUpsertSql = "
+            INSERT INTO metrics_counter_submission_daily (load_id, context_id, submission_id, date, metric_investigations, metric_investigations_unique, metric_requests, metric_requests_unique)
+            SELECT * FROM (SELECT load_id, context_id, submission_id, DATE(date) as date, count(*) as metric, 0 as metric_investigations_unique, 0 as metric_requests, 0 as metric_requests_unique
+                FROM {$this->table}
+                WHERE load_id = ? AND submission_id IS NOT NULL
+                GROUP BY load_id, context_id, submission_id, DATE(date)) AS t
+            ";
+        if (substr(Config::getVar('database', 'driver'), 0, strlen('postgres')) === 'postgres') {
+            $metricInvestigationsUpsertSql .= '
+                ON CONFLICT ON CONSTRAINT metrics_submission_daily_uc_load_id_context_id_submission_id_date DO UPDATE
+                SET metric_investigations = excluded.metric_investigations;
+                ';
+        } else {
+            $metricInvestigationsUpsertSql .= '
+                ON DUPLICATE KEY UPDATE metric_investigations = metric;
+            ';
+        }
+        // load metric_investigations
+        DB::statement($metricInvestigationsUpsertSql, [$loadId]);
+
+        // construct metric_requests upsert
+        $metricRequestsUpsertSql = "
+            INSERT INTO metrics_counter_submission_daily (load_id, context_id, submission_id, date, metric_investigations, metric_investigations_unique, metric_requests, metric_requests_unique)
+            SELECT * FROM (SELECT load_id, context_id, submission_id, DATE(date) as date, 0 as metric_investigations, 0 as metric_investigations_unique, count(*) as metric, 0 as metric_requests_unique
+                FROM {$this->table}
+                WHERE load_id = ? AND assoc_type = ?
+                GROUP BY load_id, context_id, submission_id, DATE(date)) AS t
+            ";
+        if (substr(Config::getVar('database', 'driver'), 0, strlen('postgres')) === 'postgres') {
+            $metricRequestsUpsertSql .= '
+                ON CONFLICT ON CONSTRAINT metrics_submission_daily_uc_load_id_context_id_submission_id_date DO UPDATE
+                SET metric_requests = excluded.metric_requests;
+                ';
+        } else {
+            $metricRequestsUpsertSql .= '
+                ON DUPLICATE KEY UPDATE metric_requests = metric;
+            ';
+        }
+        // load metric_requests
+        DB::statement($metricRequestsUpsertSql, [$loadId, Application::ASSOC_TYPE_SUBMISSION_FILE]);
+    }
```

Comment:
It would be better to use the Laravel toolset here, and you can use `updateOrInsert` to get compatible upsert SQL regardless of which DBMS you're using.

Reply:
I will once again try to construct it with Laravel QB, but back then I even could not properly define groupBy (with DATE(date) or similar special column constructs) if I recall it correctly... :-\

Reply:
You can use `->groupBy(DB::raw('some raw SQL clause'))` for that -- it's a really good toolset but takes some getting used to. Let me know in Slack if you get stuck!

Reply:
Where I can I will use insertUsing -- for insert with subquery -- but the upserts with subquery will need to stay so... As we have realized...